### PR TITLE
Add get_request_kwargs to check before requesting

### PR DESCRIPTION
**Pull request by kumarstack55**
When we use mechanicalsoup, we sometimes want to verify a request before submitting it.

If you merge this pull request, the package will be able to provide a way for the package's users to review the request.

This is my first pull request for this project. Please let me know if I'm missing anything.

### CODE REVIEW

File: mechanicalsoup/browser.py
Comment (hemberger) [inline mechanicalsoup/browser.py:244]: If the user provides `method` or `url` arguments to `**kwargs` in our API, it would currently raise a `TypeError` due to a repeated argument passed to `session.request` (I'm working on a unit test to demonstrate this now). However, in this patch, those keys would be overwritten by the `request_kwargs` dict update. It seems like it would be a little clunky to manually check for those keys, but maybe that's what's needed here to retain the expected behavior?

**Review by hemberger — CHANGES_REQUESTED** [review]
Hi, and thanks for your contribution to MechanicalSoup!

In principle, I think this is a fine change. I've recommended one small spelling error and identified a corner case that probably needs to be addressed. Once those are fixed, I will be ready to approve this. Then if the other maintainer approves as well we can merge it in. :)

File: mechanicalsoup/browser.py
Comment (kumarstack55) [reply mechanicalsoup/browser.py]: Thank you for pointing this out. It has been corrected.

File: mechanicalsoup/browser.py
Comment (kumarstack55) [reply mechanicalsoup/browser.py:244]: I hadn't thought about TypeError. Thank you for pointing that out.

It looks like we can check the key for backwards compatibility. There seems to be more than one way to implement it.

Which of the following implementation methods do you prefer?

(a) Raise TypeError in the package
(b) Or create a class method, e.g. `_get_request_kwargs()` that only merges the kwargs so that Python raises a TypeError.

Comment (kumarstack55) [reply mechanicalsoup/browser.py]: Since we have difficulty thinking about the msg of TypeError(msg), I thought it might be more desirable for Python to raise a TypeError.

**Comment by kumarstack55** [comment]
The group of jobs in travisci are listed as Queued, but when I click on Details to check, it looks like all the jobs are successful.

File: mechanicalsoup/browser.py
Comment (kumarstack55) [reply mechanicalsoup/browser.py]: 536ce4a398c88a3d4bdf26da19502619ebb2de67 Resolve this issue.

**Comment by kumarstack55** [comment]
> The group of jobs in travisci are listed as Queued, but when I click on Details to check, it looks like all the jobs are successful.

@hemberger Is there anything else I should do? :)

**Comment by moy** [comment]
Commits should be squashed and the typo should be fixed in the commits summary (I just fixed it in the PR title). We keep a clean history hence a PR should include several commits only if the change needs to be split into several commits, not as leftover of previous mistakes.

File: mechanicalsoup/browser.py
Comment (moy) [inline mechanicalsoup/browser.py:143]: `cls` is unused here, right? Why not use `@staticmethod` instead of `@classmethod` then?

File: mechanicalsoup/browser.py
Comment (moy) [inline mechanicalsoup/browser.py:145]: Shouldn't this comment be a docstring?

File: tests/test_browser.py
Comment (moy) [inline tests/test_browser.py:64]: Please use [pytest.raises](https://docs.pytest.org/en/stable/reference.html?highlight=raises#pytest.raises) to assert that an exception is raised.

File: tests/test_browser.py
Comment (moy) [inline tests/test_browser.py:79]: Likewise.

**Review by moy — CHANGES_REQUESTED** [review]
On overall, this looks good, but a few more nits to be fixed before this can be merged.

File: mechanicalsoup/browser.py
Comment (kumarstack55) [reply mechanicalsoup/browser.py:145]: Actually, this is my first time writing a docstring.

I'll be referring to other methods to write docstrings. Please let me know if you have any concerns.

File: tests/test_browser.py
Comment (kumarstack55) [reply tests/test_browser.py:64]: Thanks, I'll rewrite it using pytest.raises.

File: tests/test_browser.py
Comment (kumarstack55) [reply tests/test_browser.py:79]: Thanks, I'll rewrite it using pytest.raises.

File: mechanicalsoup/browser.py
Comment (kumarstack55) [reply mechanicalsoup/browser.py:143]: Thanks for the good point. I'll use `@staticmethod` to avoid making `cls` useless.

**Comment by kumarstack55** [comment]
Originally, this pull request was intended to implement `_get_request_kwargs()`, but the name was changed and `get_request_kwargs()` is the current name.

Therefore, please let me change the title of the Pull Request.

**Comment by kumarstack55** [comment]
> Commits should be squashed and the typo should be fixed in the commits summary (I just fixed it in the PR title).

I squashed the commits to one. :sparkles:

**Comment by moy** [comment]
On overall this looks good to me. The commit message would deserve a bit more love I think (explain why the change is needed and good).

**Review by moy — APPROVED** [review]

**Comment by kumarstack55** [comment]
> On overall this looks good to me. The commit message would deserve a bit more love I think (explain why the change is needed and good).

Let me make the commit log more understandable and lovely. I don't plan to make any changes to the code at this time.

**Comment by kumarstack55** [comment]
@hemberger @moy Let me know if there's anything else I need to do to get the pull request merged.

File: mechanicalsoup/browser.py
Comment (hemberger) [inline mechanicalsoup/browser.py:152]: Since this is a public method, it will need a docstring.

**Review by hemberger — CHANGES_REQUESTED** [review]
Overall, the changes look great. I've identified one simple change I'd like to see, but then I think this is ready to be merged. Thanks!

File: mechanicalsoup/browser.py
Comment (moy) [reply mechanicalsoup/browser.py:152]: Hmm, and this should be a @staticmethod, not @classmethod, right?

Comment (hemberger) [reply mechanicalsoup/browser.py:152]: I think they used classmethod because it calls
```python
return cls._get_request_kwargs(method, url, files=files, **kwargs)
```
However, this could be `Browser._get_request_kwargs`. I'm happy with it changed or as-is.

Comment (moy) [reply mechanicalsoup/browser.py:152]: OK, I'm fine with both too.

Comment (kumarstack55) [reply mechanicalsoup/browser.py:152]: I added a simple docstring to the public method.

**Comment by hemberger** [comment]
Thanks again for your contribution!

**Comment by kumarstack55** [comment]
Thank you for merging. :smile:

By the way, it occurs to me that this was the first time in my life that I've ever had my code reviewed.
I don't have a job writing programs. Furthermore, I had never even written a docstring before.
(I do some scraping as a hobby, so having Mechanicalsoup available would be very helpful. )

For me, all of the points you @hemberger @moy have given me are excellent. :sparkles:
I'm sure it would be very difficult for you to review my code.
Thanks to your thoughtful comments, I managed to get it merged. 

Again, thank you for the series of responses and for merging my request. :+1:

**Comment by moy** [comment]
Thanks for the feedback. In another life, I'm a teacher, and that's what I try to explain my students: code review is not "I went through it and it doesn't look too bad", but can be much more picky *and constructive* than that ;-).

**Pull request merged**
